File: project/assets/scripts/AbcNotation.js
import Constants, { KEY_DEF } from 'Constants'
export default new class
{
    GetParam(txt, param)
    {
        const check = txt.match(new RegExp(`${param}:(.+)`))
        return check && check[1]
    }

    GetUnitLength(txt)
    {
        return ~~(this.GetParam(txt, 'L') || '0').match(/\d+$/)[0]
    }

    GetTempo(txt)
    {
        const tempo = this.GetParam(txt, 'Q')
        const [beat, pace] = tempo.split('=')
        return [1 / ~~beat.match(/\d+$/)[0], ~~pace]
    }

    GetMeasure(txt, beat)
    {
        const pulse = this.GetParam(txt, 'M')
        const [n, p = 1] = pulse.split('/')
        return [pulse, +n / (+p * beat)]
    }

    GetVoices()
    {
        const { key, unit, voices } = this.tune
        const { accidental, keySet } = KEY_DEF[key] || {}
        this.abctxt.replace(/V:(\d+)[\x20]+(\w+)\s+([^V]*)/g, (_, id, clef, txt) =>
        {
            const unitLength = this.GetUnitLength(txt)
            let raw = this.abctxt.match(new RegExp(`V:${id}[\r\n]+([^V]+)`))[1].replace(/%\d+/g, '').trim()
            accidental && keySet.forEach(note => (raw = raw.replace(new RegExp(`${note}`, 'g'), accidental + note)))
            voices[id] = { clef, raw, unit: unitLength == 0 ? unit : 1 / unitLength }
            this.tune.unit = Math.min(this.tune.unit, voices[id].unit)
        })
    }

    GetConstants()
    {
        this.abctxt.replace(/%%(\w+)\s+([0-9\.\-]+)/g, (match, key, value) => (Constants[key] = +value))
    }

    Parse(abctxt)
    {
        this.abctxt = abctxt.replace(/\/(?!\d)/g, '/2')
        const [beat, pace] = this.GetTempo(this.abctxt)
        const [pulse, measure] = this.GetMeasure(this.abctxt, beat)
        this.tune =
        {
            index: this.GetParam(this.abctxt, 'X'),
            title: this.GetParam(this.abctxt, 'T'),
            composer: this.GetParam(this.abctxt, 'C'),
            unit: 1 / this.GetUnitLength(this.abctxt),
            key: this.GetParam(this.abctxt, 'K'),
            measure,
            pulse,
            beat,
            pace,
            voices: {},
        }
        this.GetConstants()
        this.GetVoices()
        return new Promise(resolve => resolve(this.tune))
    }
}

File: Assets/Scripts/AbcNotation.cs
using System;
using System.Linq;
using System.Collections;
using System.Collections.Generic;
using System.Text.RegularExpressions;

public class AbcNotation
{
    string index;
    string title;
    string composer;
    double unit;
    string key;
    string signature;
    float measure;
    float beat;
    float pace; // time to play a beat in seconds

    string GetParam(string txt, string param)
    {
        Match check = Regex.Match(txt, param + @":(.+)");
        return check.Groups[1].ToString().Trim();
    }

    double GetUnitLength(string txt)
    {
        var length = this.GetParam(txt, "L");
        if (length == string.Empty)
            return double.NaN;
        return 1f / Convert.ToInt32(Regex.Match(length, @"\d+$").Groups[0].ToString());
    }

    (float, int) GetTempo(string txt)
    {
        var tempo = this.GetParam(txt, "Q").Split('=');
        var piece = 1f / Convert.ToInt32(Regex.Match(tempo[0], @"\d+$").Groups[0].ToString());
        var pace = Convert.ToInt32(tempo[1]);
        return (piece, pace);
    }

    (string signature, int measure, float beat, float pace) GetMeasure(string txt)
    {
        (float piece, int pace) = this.GetTempo(txt);
        var signature = this.GetParam(txt, "M");
        var measure = signature.Split('/');
        var beat = 1f / Convert.ToInt32(measure[1]);
        return (signature, Convert.ToInt32(measure[0]), beat, 60 * piece / (beat * pace));
    }

    public string Info()
    {
        return
            $"\nindex: {this.index}" +
            $"\ntitle: {this.title}" +
            $"\ncomposer: {this.composer}" +
            $"\nunit: {this.unit}" +
            $"\nkey: {this.key}" +
            $"\nsignature: {this.signature}" +
            $"\nmeasure: {this.measure}" +
            $"\nbeat: {this.beat}" +
            $"\npace: {this.pace}"
        ;
    }

    public static AbcNotation Parse(string abctxt)
    {
        AbcNotation abc = new AbcNotation();
        abctxt = new Regex(@"\/(?!\d)").Replace(abctxt, "/2");
        (string signature, int measure, float beat, float pace) = abc.GetMeasure(abctxt);
        abc.index = abc.GetParam(abctxt, "X");
        abc.title = abc.GetParam(abctxt, "T");
        abc.composer = abc.GetParam(abctxt, "C");
        abc.unit = abc.GetUnitLength(abctxt);
        abc.GetUnitLength(abctxt);
        abc.key = abc.GetParam(abctxt, "K");
        abc.signature = signature;
        abc.measure = measure;
        abc.beat = beat;
        abc.pace = pace;
        return abc;
    }
}

